File: src/1161. Maximum Level Sum of a Binary Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @param {TreeNode} root
 * @return {number}
 */

function maxLevelSum(root) {
  let queue = [],
    level = 1,
    result = 1,
    globalMax = 0
  queue.push([root])
  while (queue.length) {
    let nodeList = queue.shift(),
      localMax = 0,
      nextList = []
    for (let node of nodeList) {
      localMax += node.val
      if (node.left) nextList.push(node.left)
      if (node.right) nextList.push(node.right)
    }
    if (localMax > globalMax) {
      result = level
      globalMax = localMax
    }
    level++
    if (nextList.length) queue.push(nextList)
  }
  return result
}

File: src/1161. Maximum Level Sum of a Binary Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int maxLevelSum(TreeNode* root) {
        queue<TreeNode*> q;
        q.push(root);
        int level = 1, maxSum = -INT_MAX, ans = 0;
        while (!q.empty()) {
            int localSum = 0;
            for (int i = q.size(); i > 0; --i) {
                auto node = q.front();
                q.pop();
                localSum += node->val;
                if (node->left) {
                    q.push(node->left);
                }
                if (node->right) {
                    q.push(node->right);
                }
            }
            if (localSum > maxSum) {
                maxSum = localSum;
                ans = level;
            }
            level++;
        }
        return ans;
    }
};